### PR TITLE
feat: Creates store to manage Portals API and data

### PR DESCRIPTION
**Pull request by nithindavid**
Fixes https://github.com/chatwoot/product/issues/489

Creates the vue store design for the new portals.

This follows the https://redux.js.org/usage/structuring-reducers/normalizing-state-shape

### CODE REVIEW

File: app/javascript/dashboard/store/modules/helpCenterPortals/getters.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  helpCenterById: (...getterArguments) => helpCenterId => {
+    const [_state] = getterArguments;
+    const helpCenter = _state.helpCenters.byId[helpCenterId];
+    if (!helpCenter) return undefined;
+
+    return {
+      ...helpCenter,
+    };
+  },
+  allHelpCenters: (...getterArguments) => {
```
Comment (muhsin-k) [inline app/javascript/dashboard/store/modules/helpCenterPortals/getters.js:18]: Spec is missing for the getter `allHelpCenters`

File: app/javascript/dashboard/store/modules/helpCenterPortals/actions.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+  fetchAllHelpCenters: async ({ commit }) => {
+    try {
+      commit('setUIFlag', { isFetching: true });
+      const { data } = await HelpCenterPortalsAPI.get();
+      data.forEach(helpCenter => {
+        const { id: helpCenterId } = helpCenter;
+
+        commit('addHelpCenterEntry', helpCenter);
+        commit('addHelpCenterId', helpCenterId);
+        commit('setHelpCenterUIFlag', {
```
Comment (pranavrajs) [inline app/javascript/dashboard/store/modules/helpCenterPortals/actions.js:13]: Please use constants for the method names here.

File: app/javascript/dashboard/store/modules/helpCenterPortals/actions.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+export const actions = {
+  fetchAllHelpCenters: async ({ commit }) => {
+    try {
+      commit('setUIFlag', { isFetching: true });
+      const { data } = await HelpCenterPortalsAPI.get();
+      data.forEach(helpCenter => {
+        const { id: helpCenterId } = helpCenter;
+
+        commit('addHelpCenterEntry', helpCenter);
```
Comment (pranavrajs) [inline app/javascript/dashboard/store/modules/helpCenterPortals/actions.js:11]: Let us not use the term help center. This is Portals API. Let us rename the variable to Portals.

File: app/javascript/dashboard/api/helpCenter/portals.js [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import ApiClient from '../ApiClient';
+
+class HelpCenterPortalsAPI extends ApiClient {
```
Comment (pranavrajs) [inline app/javascript/dashboard/api/helpCenter/portals.js:3]: Please rename this to PortalsAPI.

File: app/javascript/dashboard/store/modules/helpCenterPortals/getters.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+export const getters = {
+  uiFlagsIn: _state => helpCenterId => {
```
Comment (pranavrajs) [inline app/javascript/dashboard/store/modules/helpCenterPortals/getters.js:2]: Please use state. There is no need to use `_state`. `_` are generally used for private methods in js.

File: app/javascript/dashboard/store/modules/helpCenterPortals/getters.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  },
+  allHelpCenters: (...getterArguments) => {
+    const [_state, _getters] = getterArguments;
+
+    const helpCenters = _state.helpCenters.allIds.map(id => {
+      return _getters.helpCenterById(id);
+    });
+    return helpCenters;
+  },
+  totalHelpCentersCount: _state => _state.helpCenters.allIds.length || 0,
```
Comment (pranavrajs) [inline app/javascript/dashboard/store/modules/helpCenterPortals/getters.js:26]: Just count is enough, right? `portals/count` is readable right?

File: app/javascript/dashboard/store/modules/helpCenterPortals/actions.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
+import HelpCenterPortalsAPI from 'dashboard/api/helpCenter/portals.js';
+
+export const actions = {
+  fetchAllHelpCenters: async ({ commit }) => {
```
Comment (pranavrajs) [inline app/javascript/dashboard/store/modules/helpCenterPortals/actions.js:4]: Let us follow the naming convention here. https://discord.com/channels/897869226579222540/999557432747503698/999557936294678539

File: app/javascript/dashboard/store/modules/helpCenterPortals/actions.js [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+      },
+      helpCenterId,
+    });
+    try {
+      await HelpCenterPortalsAPI.delete(helpCenterId);
+
+      commit('removeHelpCenterEntry', helpCenterId);
+      commit('removeHelpCenterId', helpCenterId);
+    } catch (error) {
+      throw new Error(error);
```
Comment (pranavrajs) [inline app/javascript/dashboard/store/modules/helpCenterPortals/actions.js:80]: Please use the helper defined here https://github.com/chatwoot/chatwoot/blob/c0249a1b5b7ef83944d59a55e88dad8eab7b2ad3/app/javascript/dashboard/store/utils/api.js#L49-L57

Comment (pranavrajs) [inline app/javascript/dashboard/store/modules/helpCenterPortals/actions.js:80]: Please use the error helper defined here. 
https://github.com/chatwoot/chatwoot/blob/61055672384cb850415ebae70d7ccdc761bf3555/app/javascript/dashboard/store/utils/api.js#L59-L62

Comment (pranavrajs) [reply app/javascript/dashboard/store/modules/helpCenterPortals/actions.js:80]: Please update this on all the actions.

File: app/javascript/dashboard/store/modules/helpCenterPortals/actions.js [hunk 9]
```diff
@@ -0,0 +1,90 @@
+import HelpCenterPortalsAPI from 'dashboard/api/helpCenter/portals.js';
+
+export const actions = {
+  fetchAllHelpCenters: async ({ commit }) => {
+    try {
+      commit('setUIFlag', { isFetching: true });
+      const { data } = await HelpCenterPortalsAPI.get();
+      data.forEach(helpCenter => {
```
Comment (pranavrajs) [inline app/javascript/dashboard/store/modules/helpCenterPortals/actions.js:8]: Wouldn't this trigger multiple renders in the UI? Can we do it in bulk?

File: app/javascript/dashboard/store/modules/helpCenterPortals/getters.js (hunk 5 above)
Comment (pranavrajs) [reply app/javascript/dashboard/store/modules/helpCenterPortals/getters.js:2]: Please update the names in the entire file.

File: app/javascript/dashboard/store/modules/helpCenterPortals/getters.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+  isFetchingHelpCenters: _state => _state.uiFlags.isFetching,
+  helpCenterById: (...getterArguments) => helpCenterId => {
+    const [_state] = getterArguments;
+    const helpCenter = _state.helpCenters.byId[helpCenterId];
+    if (!helpCenter) return undefined;
+
+    return {
+      ...helpCenter,
+    };
```
Comment (pranavrajs) [inline app/javascript/dashboard/store/modules/helpCenterPortals/getters.js:16]: Can't we just return the object? This check and returning a new object seem unnecessary.

File: app/javascript/dashboard/store/modules/helpCenterPortals/mutations.js [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    });
+  },
+
+  addHelpCenterId($state, helpCenterId) {
+    $state.helpCenters.allIds.push(helpCenterId);
+  },
+
+  updateHelpCenterEntry($state, helpCenter) {
+    const helpCenterId = helpCenter.id;
+    if (!helpCenterId) return;
```
Comment (pranavrajs) [inline app/javascript/dashboard/store/modules/helpCenterPortals/mutations.js:26]: Why do we need this check? Sending an update action without an id might not be done anywhere. Even if it was sent, the next line would catch it.

File: app/javascript/dashboard/store/modules/helpCenterPortals/mutations.js [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+    if (!$state.helpCenters.allIds.includes(helpCenterId)) return;
+
+    Vue.set($state.helpCenters.byId, helpCenterId, {
+      ...helpCenter,
+    });
+  },
+
+  removeHelpCenterEntry($state, helpCenterId) {
+    if (!helpCenterId) return;
```
Comment (pranavrajs) [inline app/javascript/dashboard/store/modules/helpCenterPortals/mutations.js:35]: This check seems unnecessary (Similar to the previous comment)

**Review by pranavrajs — CHANGES_REQUESTED** [review]
Please check the individual comments.

File: app/javascript/dashboard/store/modules/helpCenterPortals/actions.js [hunk 13]
```diff
@@ -0,0 +1,82 @@
+import HelpCenterPortalsAPI from 'dashboard/api/helpCenter/portals.js';
```
Comment (muhsin-k) [inline app/javascript/dashboard/store/modules/helpCenterPortals/actions.js:1]:
```suggestion
import PortalsAPI from 'dashboard/api/helpCenter/portals.js';
```

File: app/javascript/dashboard/store/modules/helpCenterPortals/actions.js [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+import { throwErrorMessage } from 'dashboard/store/utils/api';
+import { types } from './mutations';
+
+export const actions = {
+  fetchAllPortals: async ({ commit }) => {
+    try {
+      commit(types.SET_UI_FLAG, { isFetching: true });
+      const { data } = await HelpCenterPortalsAPI.get();
```
Comment (muhsin-k) [inline app/javascript/dashboard/store/modules/helpCenterPortals/actions.js:9]:
```suggestion
      const { data } = await PortalsAPI.get();
```

File: app/javascript/dashboard/store/modules/helpCenterPortals/actions.js [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+      throwErrorMessage(error);
+    } finally {
+      commit(types.SET_UI_FLAG, { isFetching: false });
+    }
+  },
+
+  createPortal: async ({ commit }, params) => {
+    commit(types.SET_UI_FLAG, { isCreating: true });
+    try {
+      const { data } = await HelpCenterPortalsAPI.create(params);
```
Comment (muhsin-k) [inline app/javascript/dashboard/store/modules/helpCenterPortals/actions.js:24]:
```suggestion
      const { data } = await PortalsAPI.create(params);
```

File: app/javascript/dashboard/store/modules/helpCenterPortals/actions.js [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+  updatePortal: async ({ commit }, { helpCenter }) => {
+    const helpCenterId = helpCenter.id;
+    commit(types.SET_HELP_PORTAL_UI_FLAG, {
+      uiFlags: {
+        isUpdating: true,
+      },
+      helpCenterId,
+    });
+    try {
+      const { data } = await HelpCenterPortalsAPI.update(helpCenter);
```
Comment (muhsin-k) [inline app/javascript/dashboard/store/modules/helpCenterPortals/actions.js:45]:
```suggestion
      const { data } = await PortalsAPI.update(helpCenter);
```

File: app/javascript/dashboard/store/modules/helpCenterPortals/actions.js [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+
+  deletePortal: async ({ commit }, { helpCenterId }) => {
+    commit(types.SET_HELP_PORTAL_UI_FLAG, {
+      uiFlags: {
+        isDeleting: true,
+      },
+      helpCenterId,
+    });
+    try {
+      await HelpCenterPortalsAPI.delete(helpCenterId);
```
Comment (muhsin-k) [inline app/javascript/dashboard/store/modules/helpCenterPortals/actions.js:67]:
```suggestion
      await PortalsAPI.delete(helpCenterId);
```

**Review by muhsin-k — APPROVED** [review]
Added a minor comment. PR looks good to me

File: app/javascript/dashboard/store/modules/helpCenterPortals/actions.js [hunk 18]
```diff
@@ -0,0 +1,82 @@
+import HelpCenterPortalsAPI from 'dashboard/api/helpCenter/portals.js';
+import { throwErrorMessage } from 'dashboard/store/utils/api';
+import { types } from './mutations';
+
+export const actions = {
+  fetchAllPortals: async ({ commit }) => {
```
Comment (pranavrajs) [inline app/javascript/dashboard/store/modules/helpCenterPortals/actions.js:6]:
```suggestion
  index: async ({ commit }) => {
```

File: app/javascript/dashboard/store/modules/helpCenterPortals/actions.js [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+      commit(types.ADD_MANY_PORTALS_ENTRY, data);
+      commit(types.ADD_MANY_PORTALS_IDS, portalIds);
+    } catch (error) {
+      throwErrorMessage(error);
+    } finally {
+      commit(types.SET_UI_FLAG, { isFetching: false });
+    }
+  },
+
+  createPortal: async ({ commit }, params) => {
```
Comment (pranavrajs) [inline app/javascript/dashboard/store/modules/helpCenterPortals/actions.js:21]:
```suggestion
  create: async ({ commit }, params) => {
```

File: app/javascript/dashboard/store/modules/helpCenterPortals/actions.js [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+      commit(types.ADD_PORTAL_ENTRY, data);
+      commit(types.ADD_PORTAL_ID, helpCenterId);
+    } catch (error) {
+      throwErrorMessage(error);
+    } finally {
+      commit(types.SET_UI_FLAG, { isCreating: false });
+    }
+  },
+
+  updatePortal: async ({ commit }, { helpCenter }) => {
```
Comment (pranavrajs) [inline app/javascript/dashboard/store/modules/helpCenterPortals/actions.js:36]:
```suggestion
  update: async ({ commit }, { helpCenter }) => {
```

File: app/javascript/dashboard/store/modules/helpCenterPortals/actions.js [hunk 21]
```diff
@@ -0,0 +1,82 @@
...
+      commit(types.SET_HELP_PORTAL_UI_FLAG, {
+        uiFlags: {
+          isUpdating: false,
+        },
+        helpCenterId,
+      });
+    }
+  },
+
+  deletePortal: async ({ commit }, { helpCenterId }) => {
```
Comment (pranavrajs) [inline app/javascript/dashboard/store/modules/helpCenterPortals/actions.js:59]:
```suggestion
  delete: async ({ commit }, { helpCenterId }) => {
```

File: app/javascript/dashboard/store/modules/helpCenterPortals/actions.js [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+    } finally {
+      commit(types.SET_UI_FLAG, { isFetching: false });
+    }
+  },
+
+  createPortal: async ({ commit }, params) => {
+    commit(types.SET_UI_FLAG, { isCreating: true });
+    try {
+      const { data } = await HelpCenterPortalsAPI.create(params);
+      const { id: helpCenterId } = data;
```
Comment (pranavrajs) [inline app/javascript/dashboard/store/modules/helpCenterPortals/actions.js:25]: portalId

**Review by pranavrajs — CHANGES_REQUESTED** [review]
@nithindavid There are references of helpCenter instead of portal. Please update those. Also, please update the actions to follow the naming convention. https://chatwoot.com/docs/handbook/engineering/frontend-guidelines#avoid-manipulating-dom-directly

**Pull request merged**